haystack/document_store/milvus.py

```diff
@@ -0,0 +1,384 @@
+import logging
+from typing import Any, Dict, Generator, List, Optional, Union
+import numpy as np
+
+from milvus import IndexType, MetricType, Milvus
+from scipy.special import expit
+from tqdm import tqdm
+
+from haystack import Document
+from haystack.document_store.sql import SQLDocumentStore
+from haystack.retriever.base import BaseRetriever
+from haystack.utils import get_batches_from_generator
+
+logger = logging.getLogger(__name__)
+
+
+class MilvusDocumentStore(SQLDocumentStore):
+    """
+    Document store for very large scale embedding based dense retrievers like the DPR.
+    It implements the Milvus (https://github.com/milvus-io/milvus) to perform similarity search on vectors.
+    The document text and meta-data (for filtering) is stored using the SQLDocumentStore, while the vector embeddings
+    are indexed in a Milvus Server Index. Refer https://milvus.io/docs/v0.10.5/tuning.md for performance tuning option
+    """
+
+    def __init__(
+            self,
+            sql_url: str = "sqlite:///",
+            server_uri: str = "tcp://localhost:19530",
+            connection_pool: str = "SingletonThread",
+            index: str = "document",
+            vector_dim: int = 768,
+            index_file_size: int = 1024,
+            metric_type: MetricType = MetricType.IP,
```

Comment:
In milvus terminology the metric is the function to calculate distance/similarity between two vecs. In the other documentstores in Haystack we already have a similar parameter to define this, it's called `similarity` . See here: https://github.com/deepset-ai/haystack/blob/master/haystack/document_store/base.py#L19 
Let's keep the naming and type (str) consistent. This will make it easier for users to understand when they switch between doc stores.

Reply:
Milvus doesn't support a "native" cosine similarity metric, but rather expect the users to normalize the vectors before indexing (then inner product == cos. similarity) (see https://milvus.io/docs/v0.10.0/metric.md#Inner-product-IP). For DPR IP is fine. For some embedding retrievers (e.g. sentence transformers) cosine similarity is preferred. Maybe we can add a "normalization" mechanism here ourselves in a later PR